concrete-core/src/crypto/secret/glwe.rs

```diff
@@ -89,35 +188,171 @@ impl<Cont> GlweSecretKey<Cont> {
     /// assert_eq!(secret_key.key_size(), GlweDimension(256));
     /// assert_eq!(secret_key.polynomial_size(), PolynomialSize(11));
     /// ```
-    pub fn from_container(cont: Cont, poly_size: PolynomialSize) -> Self
+    pub fn binary_from_container(cont: Cont, poly_size: PolynomialSize) -> Self
     where
         Cont: AsRefSlice,
     {
         ck_dim_div!(cont.as_slice().len() => poly_size.0);
         GlweSecretKey {
             tensor: Tensor::from_container(cont),
             poly_size,
+            kind: PhantomData,
         }
     }
+}
 
+impl<Cont> GlweSecretKey<TernaryKeyKind, Cont> {
+    /// Creates a ternary key from a container.
+    ///
+    /// # Notes
+    ///
+    /// This method does not fill the container with random data. It merely wraps the container in
+    /// the appropriate type. For a method that generate a new random key see
+    /// [`GlweSecretKey::generate`].
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use concrete_core::crypto::secret::*;
+    /// use concrete_core::crypto::*;
+    /// use concrete_core::math::polynomial::PolynomialSize;
+    /// let secret_key =
+    ///     GlweSecretKey::ternary_from_container(vec![0 as u8; 11 * 256], PolynomialSize(11));
+    /// assert_eq!(secret_key.key_size(), GlweDimension(256));
+    /// assert_eq!(secret_key.polynomial_size(), PolynomialSize(11));
+    /// ```
+    pub fn ternary_from_container(cont: Cont, poly_size: PolynomialSize) -> Self
+    where
+        Cont: AsRefSlice,
+    {
+        ck_dim_div!(cont.as_slice().len() => poly_size.0);
+        GlweSecretKey {
+            tensor: Tensor::from_container(cont),
+            poly_size,
+            kind: PhantomData,
+        }
+    }
+}
+
+impl<Cont> GlweSecretKey<GaussianKeyKind, Cont> {
+    /// Creates a gaussian key from a container.
+    ///
+    /// # Notes
+    ///
+    /// This method does not fill the container with random data. It merely wraps the container in
+    /// the appropriate type. For a method that generate a new random key see
+    /// [`GlweSecretKey::generate`].
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use concrete_core::crypto::secret::*;
+    /// use concrete_core::crypto::*;
+    /// use concrete_core::math::polynomial::PolynomialSize;
+    /// let secret_key =
+    ///     GlweSecretKey::binary_from_container(vec![0 as u8; 11 * 256], PolynomialSize(11));
+    /// assert_eq!(secret_key.key_size(), GlweDimension(256));
+    /// assert_eq!(secret_key.polynomial_size(), PolynomialSize(11));
+    /// ```
+    pub fn gaussian_from_container(cont: Cont, poly_size: PolynomialSize) -> Self
+    where
+        Cont: AsRefSlice,
+    {
+        ck_dim_div!(cont.as_slice().len() => poly_size.0);
+        GlweSecretKey {
+            tensor: Tensor::from_container(cont),
+            poly_size,
+            kind: PhantomData,
+        }
+    }
+}
+
+impl<Cont> GlweSecretKey<UniformKeyKind, Cont> {
```

Comment:
What would be a uniform key? I thought we could only have uniform binary, uniform ternary or gaussian?

Reply:
Yes, I mistakenly thought we could have some. My bad !

Reply:
Actually they do exist !